kani-compiler/src/kani_middle/transform/check_aliasing/instrumentation.rs

```diff
@@ -0,0 +1,446 @@
+    /// locals in the body that can be used to call them
+    fn_pointers: HashMap<MirInstance, Local>,
+    /// The span of the body
+    span: Span,
+    /// The minimum processed instruction.
+    /// All instructions before this one belong to the original
+    /// source code, and have not yet been analyzed.
+    min_processed: SourceInstruction,
+    /// The index after which "ghost" instrumentation code
+    /// may be added.
+    ghost_index: SourceInstruction,
+    /// The body being instrumented
+    body: MutableBody,
+}
+
+impl<'tcx, 'cache> InstrumentationData<'tcx, 'cache> {
+    /// Move bb0 to the end to start instrumentation there.
+    fn prepare_body(body: Body) -> MutableBody {
+        let mut body = MutableBody::from(body);
+        let span;
+        let mut source;
+        match body.blocks()[0].statements.len() {
+            0 => {
+                source = SourceInstruction::Terminator { bb: 0 };
+                span = body.blocks()[0].terminator.span;
+            }
+            _ => {
+                source = SourceInstruction::Statement { idx: 0, bb: 0 };
+                span = body.blocks()[0].terminator.span;
+            }
+        }
+        let kind = TerminatorKind::Goto { target: body.blocks().len() };
+        let terminator = Terminator { kind, span };
+        body.insert_terminator(&mut source, InsertPosition::Before, terminator);
+        body
+    }
+
+    /// Using a (potentially) pre-populated cache of resolved generic
+    /// functions, and the StableMir body "body", initialize the instrumentation
+    /// pass data.
+    pub fn new(tcx: TyCtxt<'tcx>, cache: &'cache mut Cache, body: Body) -> Self {
+        let span = body.span;
+        let mut body = Self::prepare_body(body);
+        let meta_stack = HashMap::new();
+        let local_count = body.locals().len();
+        let fn_pointers = HashMap::new();
+        let unit = body.new_local(Ty::new_tuple(&[]), span, Mutability::Not);
+        let valid = body.new_local(
+            Ty::from_rigid_kind(stable_mir::ty::RigidTy::Bool),
+            span,
+            Mutability::Mut,
+        );
+        let bb = body.blocks().len() - 1;
+        let min_processed = match body.blocks()[bb].statements.len() {
+            0 => SourceInstruction::Terminator { bb },
+            n => SourceInstruction::Statement { idx: n - 1, bb },
+        };
+        let ghost_index = SourceInstruction::Terminator { bb: 0 };
+        InstrumentationData {
+            tcx,
+            cache,
+            meta_stack,
+            local_count,
+            unit,
+            valid,
+            fn_pointers,
+            span,
+            min_processed,
+            ghost_index,
+            body,
+        }
+    }
+
+    /// Register the function described by the diagnostic
+    /// and generic arguments in "Signature".
+    fn register_fn(&mut self, callee: Signature) -> Result<Local> {
+        let cache = &mut self.cache;
```

Comment:
You don't have to introduce extra variables: e.g.: calling `self.cache.register(...)` is good, too